### ISSUE TITLE
API for batched input?

### ISSUE THREAD

**Issue by 0x1997** [turboderp/exllama#50]
Thanks for this great project. The inference speed is exceptional. However it seems the generator api only supports single string input. When serving concurrent requests, batching of inputs will be needed for better thoughput.

**Comment by turboderp** [comment]
The generator only takes single strings, but the model should be able to process batches. It's kind of untested, though, and I never got around to deciding on a good API batch generation. It gets a little complicated with stop conditions and padding and all that. I always intended to support batches, but I'll add it to the todo list so I don't forget.

**Comment by ri938** [comment]
I did try to pass identical combination of tokens into model.forward (using batch size 2) and the logits returned didnt not match for each row of the batch. They were close but not identicaly. The probabilities of the most likely next token in both cases was about 1% different.

**Comment by turboderp** [comment]
It's probably a broadcasting failure somewhere. There's basically no difference between processing two batches of 100 tokens vs one batch of 200 tokens, except keeping the tensor shapes in order. So it's definitely fixable. And on the generator side it just comes down to padding. So I'll get to it, just as soon as I finish adding LoRA support.

**Comment by turboderp** [comment]
With the latest commit `generate_simple()` can now accept a list of prompts, complete them as a batch and output a list of completions. See `example_batch.py`. Note that the cache needs to be allocated with an appropriate batch size, and of course VRAM usage goes up accordingly.

**Comment by juliensalinas** [comment]
Thank you for the amazing work!

I am getting perfect results with `generate_simple()` but when used with batch size > 1 the results are gibberish.

Here is an example that works perfectly with batch size = 1, using `example_basic.py`:
prompt = "I will tell you a very long story:"
output --> "I will tell you a very long story:
- I was in the middle of my PhD studies at that time. My supervisor was not happy with me, because he wanted to see some results from my research and I had none. So we agreed on a plan for the next 6 months. He would help me to find a job (which is difficult when you are a foreigner) and"

Here is another example that fails with batch size = 2, using `example_batch.py` (the `batch_size` is properly set in `ExLlamaCache()`):
prompts = ["I will tell you a very long story:", "Hi"]
output --> ['I will tell you a very long story:\n. . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .', 'Hippus, J. H. Hippus, 1930, 2648, 2735, 2736, 2737, 2738, 2739, 2740, 2741, 2742, 2743, 2744, 2745, 2746, 2747, 2748, 2749, 2750, 2751, 2752, 2753, 2754, 2755, 2756, 2757, 2758, 2759, 2760, 2761, 2762, 2763, 27']

Is it correctly working on your end?

Thank you so much.

**Comment by turboderp** [comment]
I'm not sure what would cause that. The `example_batch.py` script seems to run fine here, even with `prompts = ["I will tell you a very long story:", "Hi"]` Could it be something with the model you're using or the sampling parameters? If you pass a single prompt without changing anything else in `example_batch.py`, does it work?

**Comment by juliensalinas** [comment]
Thank you @turboderp , let me test the above and get back to you asap 👍🏻

**Comment by juliensalinas** [comment]
Hello again @turboderp ,
I made some more tests and here are my conclusions:

- Your [batch example](https://github.com/turboderp/exllama/blob/master/example_batch.py) works correctly
- I still get gibberish outputs with my 2 above examples (it's not occurring every time though)
- The sampling parameters don't seem to be the cause

So my conclusion for the moment is that there is **no** problem with ExLLaMA.

For my tests I used AutoGPTQ and quantized LLaMA 7B in 4 bits. So maybe this is simply an effect of the 4-bit quantization on the 7B version. I will make some more tests with the 30B and 60B versions and see how it goes.

Thanks again!
